jobsdb/unionQuery.go

```diff
@@ -0,0 +1,284 @@
+	jd.assertError(err)
+
+	for rows.Next() {
+		var customer string
+		var count int
+		err := rows.Scan(&customer, &count)
+		jd.assertError(err)
+		unionMap[customer] = count
+	}
+
+	return unionMap
+}
+
+//Unprocessed
+
+func (jd *HandleT) getUnprocessedUnionQuerystring(unionMap map[string]int, ds dataSetT, params GetQueryParamsT) string {
+	var queries []string
+
+	for customer, count := range unionMap {
+		//do cache stuff here
+		queries = append(queries, jd.getSingleCustomerUnprocessedQueryString(customer, count, ds, params, true))
+	}
+
+	return `(` + strings.Join(queries, `) UNION (`) + `)`
+}
+
+func (jd *HandleT) getSingleCustomerUnprocessedQueryString(customer string, count int, ds dataSetT, params GetQueryParamsT, order bool) string {
+	customValFilters := params.CustomValFilters
+	parameterFilters := params.ParameterFilters
+	var sqlStatement string
+
+	if useJoinForUnprocessed {
+		// event_count default 1, number of items in payload
+		sqlStatement = fmt.Sprintf(
+			selectQuery+
+				`FROM %[1]s AS jobs `+
+				`LEFT JOIN %[2]s AS job_status ON jobs.job_id=job_status.job_id `+
+				`WHERE job_status.job_id is NULL AND customer='%[3]s'`,
+			ds.JobTable, ds.JobStatusTable, customer)
+	} else {
+		sqlStatement = fmt.Sprintf(
+			selectQuery+
+				` FROM %[1]s AS jobs `+
+				`WHERE customer='%[3]s' AND jobs.job_id NOT IN (SELECT DISTINCT(job_status.job_id) FROM %[2]s AS job_status)`,
+			ds.JobTable, ds.JobStatusTable, customer)
+	}
+
+	if len(customValFilters) > 0 && !params.IgnoreCustomValFiltersInQuery {
+		sqlStatement += " AND " + constructQuery(jd, "jobs.custom_val", customValFilters, "OR")
+	}
+
+	if len(parameterFilters) > 0 {
+		sqlStatement += " AND " + constructParameterJSONQuery("jobs", parameterFilters)
+	}
+
+	//avoinding AfterJobID for now
+
+	if params.UseTimeFilter {
+		sqlStatement += fmt.Sprintf(" AND created_at < %s", params.Before)
+	}
+
+	if order {
+		sqlStatement += " ORDER BY jobs.job_id"
+	}
+
+	if count > 0 {
+		sqlStatement += fmt.Sprintf(" LIMIT %d", count)
+	}
+
+	if params.EventCount > 0 {
+		sqlStatement = fmt.Sprintf(`SELECT * FROM (`+sqlStatement+`) AS subquery WHERE running_event_counts - event_count + 1 <= %d;`, params.EventCount)
+	}
+
+	return sqlStatement
+}
+
+func (jd *HandleT) GetUnprocessedUnion(unionMap map[string]int, params GetQueryParamsT) []*JobT {
+	// if params.JobCount == 0 { //how do we use this check..? is it required anymore
+	// 	return []*JobT{}
+	// }
+
+	//add stats
+
+	//The order of lock is very important. The migrateDSLoop
+	//takes lock in this order so reversing this will cause
+	//deadlocks
+	jd.dsMigrationLock.RLock()
+	jd.dsListLock.RLock()
+	defer jd.dsMigrationLock.RUnlock()
+	defer jd.dsListLock.RUnlock()
+
+	dsList := jd.getDSList(false)
+	outJobs := make([]*JobT, 0)
+
+	//removed count assert, because that params.count is not used..?
+
+	for _, ds := range dsList {
+		jobs := jd.getUnprocessedUnionDS(ds, unionMap, params)
+		outJobs = append(outJobs, jobs...)
```

Comment:
What if all customer counts are zero? Also is it possible to get to negative counts? I think we need to double-check for that.

Early stop if nothing more to read?

```suggestion
		jobs := jd.getUnprocessedUnionDS(ds, unionMap, params)
		outJobs = append(outJobs, jobs...)
		if len(unionMap) == 0 {
		    break 
		}
```

Reply:
I was under the impression that we would always pick a minimum number of jobs per customer(the `delta` value from fairness equation). 
And the cache, which is next in line would help to not query all customers:
>  if the cache says there's no jobs, we won't include that customer in the union query.

And yes, I'll add the check for negative counts shortly.

Reply:
I just realized what you're saying. This is required, yes.